packages/material-ui/src/ButtonBase/TouchRipple.js

```diff
@@ -98,173 +98,193 @@ export const styles = theme => ({
+      // Touche devices
+      if (event.touches) {
+        // Prepare the ripple effect.
+        startTimerCommit.current = () => {
+          startCommit({ pulsate, rippleX, rippleY, rippleSize, cb });
+        };
+        // Delay the execution of the ripple effect.
+        startTimer.current = setTimeout(() => {
+          if (startTimerCommit.current) {
+            startTimerCommit.current();
+            startTimerCommit.current = null;
+          }
+        }, DELAY_RIPPLE); // We have to make a tradeoff with this value.
+      } else {
+        startCommit({ pulsate, rippleX, rippleY, rippleSize, cb });
+      }
+    },
+    [centerProp, startCommit],
+  );
 
-    this.setState(
-      state => ({
-        nextKey: state.nextKey + 1,
-        ripples: [
-          ...state.ripples,
-          <Ripple
-            key={state.nextKey}
-            classes={this.props.classes}
-            timeout={{
-              exit: DURATION,
-              enter: DURATION,
-            }}
-            pulsate={pulsate}
-            rippleX={rippleX}
-            rippleY={rippleY}
-            rippleSize={rippleSize}
-          />,
-        ],
-      }),
-      cb,
-    );
-  };
+  const pulsate = React.useCallback(() => {
+    start({}, { pulsate: true });
+  }, [start]);
 
-  stop = (event, cb) => {
-    clearTimeout(this.startTimer);
+  const stop = React.useCallback((event, cb) => {
+    clearTimeout(startTimer.current);
 
     // The touch interaction occurs too quickly.
     // We still want to show ripple effect.
-    if (event.type === 'touchend' && this.startTimerCommit) {
+    if (event.type === 'touchend' && startTimerCommit.current) {
       event.persist();
-      this.startTimerCommit();
-      this.startTimerCommit = null;
-      this.startTimer = setTimeout(() => {
-        this.stop(event, cb);
+      startTimerCommit.current();
+      startTimerCommit.current = null;
+      startTimer.current = setTimeout(() => {
+        stop(event, cb);
       });
       return;
     }
 
-    this.startTimerCommit = null;
+    startTimerCommit.current = null;
 
-    this.setState(({ ripples }) => {
-      if (ripples && ripples.length) {
-        return { ripples: ripples.slice(1) };
+    setRipples(oldRipples => {
+      if (oldRipples.length > 0) {
+        return oldRipples.slice(1);
       }
-      return null;
-    }, cb);
-  };
+      return oldRipples;
+    });
+    rippleCallback.current = cb;
+  }, []);
+
+  React.useImperativeHandle(
+    ref,
+    () => ({
+      pulsate,
+      start,
+      stop,
+    }),
+    [pulsate, start, stop],
+  );
+
+  return (
+    <span className={clsx(classes.root, className)} ref={container} {...other}>
+      <TransitionGroup component={null} enter exit>
+        {ripples}
+      </TransitionGroup>
+    </span>
+  );
+});
 
-  render() {
-    const { center, classes, className, ...other } = this.props;
+const TouchRippleMemo = React.memo(TouchRipple);
```

Comment:
The `react/prop-types` lint rule wasn't picking up the prop types with this wrapping React.forwardRef directly.